### ISSUE TITLE
[NNBD] Type promotion fails in async functions

### ISSUE THREAD

**Issue by nex3** [dart-lang/sdk#45346]
The following function:

```dart
Future<String> foo(int arg) async {
  String? value;
  if (arg > 0) {
    value = await Future.value("greater than");
  } else {
    value = "less than";
  }
  return value;
}
```

Produces this error:

```
Analyzing test.dart...
  error • A value of type 'String?' can't be returned from the function 'foo' because it has a return type of 'Future<String>'. • test
.dart:9:10 • return_of_invalid_type                                                                                                  
1 error found.
```

`value` is clearly never `null` here, and in fact if `Future.value()` is replaced by a plain string it's detected correctly.

**Comment by stereotype441** [comment]
Yeah, this is an unfortunate case.  Thanks for bringing it up and prompting me to look at it with fresh eyes.

What's happening here is that when type inference reaches `Future.value("greater than")`, it looks up the type signature of the `Future<T>.value` constructor (which is `Future<T> Function([FutureOr<T>?])`) and then does the following reasoning steps:
- The context type is `FutureOr<String?>` (meaning the value produced by the constructor must be a subtype of either `String?` or `Future<String?>` in order for the assignment to succeed).  Therefore `T` must be a subtype of `String?`.
- The argument type is `String`.  Therefore `String` must be a subtype of `FutureOr<T>`, and thus `String` must be a subtype of `T`.

It doesn't have any more information, so it has to choose a type of either `String` or `String?` for the type.  It chooses `String?`, so it's creating a Future of type `Future<String?>`.  Awaiting this future produces a `String?`, so the assignment to `value` doesn't promote `value` to `String`.

For your situation there are two solutions you could consider:
1. Change the declaration of `value` from `String? value;` to `String value;`.  This way your code won't require type promotion because `value` will have the right type from the start.  (Note that this is allowed under null safety, because flow analysis can see that `value` is always assigned before it is used, so even though `value` has a non-nullable type, you don't have to initialize it).  This is probably the best solution for your example.
2. Specify the type argument explicitly in the call to `Future.value` by changing it to `Future<String>.value`.  This will ensure that both assignments to `value` promote from `String?` to `String`.

Of course, the obvious follow up question is, could we make type inference smarter, so that it understands that `Future<String>` was the type you really wanted, rather than `Future<String?>`?  I believe the answer is yes, we could, but we would have to add some information to the type system.  Specifically, we would have to add the information that in the methods on class `Future<T>`, `T` is only ever used in a covariant position (e.g. as a return type)*, which means that a `Future<String>` is always safe to use where a `Future<String?>` is expected.  Note that this isn't true of other generic types, like `List`:
```dart
List<String?> x = List.filled(10, "s");
x[0] = null;
```
The exact same type inference situation exists here, but it's crucial that we infer `List<String?>` in this case.  (If we inferred `List<String>`, then there would be a runtime failure on the second line).

I think we could address this situation as part of the "variance" feature, since that feature allows us to supply exactly the information that's needed; in the case of `Future`, the type parameter would be an "out" parameter, so type inference would know that `Future<String>` is safe to use where `Future<String?>` is expected, whereas in the case of `List`, the type parameter would be an "inout" parameter, so type inference would know that it has to choose `List<String?>` to be safe.

@leafpetersen @kallentu does that idea seem feasible to you?

(Technically I lied when I said that `T` is only ever used in a covariant position in methods on `Future`, because `Future<T>.timeout` has a contravariant use.  It's signature is `Future<T> Function(Duration, {FutureOr<T> onTimeout})`.  I suspect we could address that problem with some additional type system cleverness, but I haven't thought the whole thing through.)

**Comment by nex3** [comment]
I didn't realize that this *only* occurred when the right-hand side of the expression could be inferred as nullable. In that case, it's easy for me to work around :smiley:.

**Comment by leafpetersen** [comment]
@stereotype441 This is in fact how inference works in the prototype that @kallentu implemented.  If you run the following code with `--enable-experiment=variance`:
```dart
class C1<out T> {
  C1(T x);
}

class C2<T> {
  C2(T x);
}

void main() {
  C1<int?> a = C1(3);
  print(a.runtimeType);
  C2<int?> b = C2(3);
  print(b.runtimeType);
}
```

You get the following output:

```
C1<int>
C2<int?>
```

In the runtime checked covariant example (`C2`), inference "pins" the result immediately based on the context type, for the reason you describe.  In the statically checked covariant example (`C1`) inference chooses the principle "most precise" type.

**Comment by stereotype441** [comment]
@leafpetersen awesome, glad to hear it!
